### PR TITLE
Starting the networking layer rewrite

### PR DESCRIPTION
**Pull request by apederson94**
Started writing an alternative networking layer to use Retrofit and OkHttp instead of a custom networking layer written using Apache HTTP client (and OkHttp for Android client).

- This will be the first PR of many
- This should contain some of the basic building blocks necessary for others to start piggybacking and helping to generate a more easy to maintain SDK

### CODE REVIEW

File: src/integration-test/java/WorkspaceResourcesIT.java [hunk 1]
```diff
@@ -52,8 +52,8 @@ void testWorkspaceMethods() throws IOException, SmartsheetException {
         testShareWorkspace();
         testGetWorkspace();
         testListWorkspaces();
-        testUpdateWorkspace();
-        testDeleteWorkspace();
+//        testUpdateWorkspace();
```
Comment (zromano) [inline src/integration-test/java/WorkspaceResourcesIT.java]: Do we want these commented out in the test?

File: src/main/java/com/smartsheet/api/internal/WorkspaceResourcesImpl.java [hunk 2]
```diff
@@ -95,12 +112,19 @@ public WorkspaceResourcesImpl(SmartsheetImpl smartsheet) {
...
         if (parameters != null) {
-            path += parameters.toQueryString();
+            queryParams = parameters.toHashMap();
+        }
+
+        try {
+            PagedResult<Workspace> result = this.service.listWorkspaces(queryParams).execute().body();
+            return result;
+        } catch (Exception e) {
+            log.info("failure calling 'listWorkspaces'", e);
```
Comment (zromano) [inline src/main/java/com/smartsheet/api/internal/WorkspaceResourcesImpl.java]: This should probably be `log.error`

File: src/main/java/com/smartsheet/api/internal/oauth/OAuthFlowImpl.java_859283488703776
Comment (zromano) [inline src/main/java/com/smartsheet/api/internal/oauth/OAuthFlowImpl.java_859283488703776:1]: Can we delete this file? 😆

File: src/main/java/com/smartsheet/api/internal/http/service/WorkspaceService.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
+package com.smartsheet.api.internal.http.service;
+
+/*
+ * #[license]
+ * Smartsheet Java SDK
+ * %%
+ * Copyright (C) 2014 - 2023 Smartsheet
```
Comment (zromano) [inline src/main/java/com/smartsheet/api/internal/http/service/WorkspaceService.java:7]: Just curious, did you just copy/paste the license to this file or set up a way to automatically add this?

I know in IntelliJ you can make it automatically add/update license info, but I've never used it.

File: src/main/java/com/smartsheet/api/internal/http/RequestAndResponseData.java [hunk 4]
```diff
@@ -303,10 +309,79 @@ public static RequestAndResponseData of(HttpRequestBase request, HttpEntitySnaps
         return new RequestAndResponseData(requestBuilder.build(), responseBuilder.build());
     }
 
+    /**
+     * factory method for creating a RequestAndResponseData object from request and response okhttp data with the specified trace fields
+     */
+    public static RequestAndResponseData of(Request request, Response response, Set<Trace> traces)
```
Comment (zromano) [inline src/main/java/com/smartsheet/api/internal/http/RequestAndResponseData.java:315]: At first glance, this method is a little scary. Is there any way we can break this up into smaller methods?

On this topic, it doesn't have to be in this MR to the feature branch, but can we add tests for all the new code?

Comment (zromano) [reply src/main/java/com/smartsheet/api/internal/http/RequestAndResponseData.java:315]: maybe buildRequest and buildResponse methods?

File: src/main/java/com/smartsheet/api/internal/http/RequestAndResponseData.java [hunk 5]
```diff
@@ -303,10 +309,79 @@ public static RequestAndResponseData of(HttpRequestBase request, HttpEntitySnaps
...
+        if (request != null) {
+            requestBuilder.withCommand(request.method() + " " + request.url());
+            boolean binaryBody = false;
+            if (traces.contains(Trace.RequestHeaders)) {
+                requestBuilder.withHeaders();
+                for (Pair<? extends String, ? extends String> header : request.headers()) {
+                    String headerName = header.getFirst();
+                    String headerValue = header.getSecond();
+                    if ("Authorization".equals(headerName) && headerValue.length() > 0) {
+                        headerValue = "Bearer ****" + headerValue.substring(Math.max(0, headerValue.length() - 4));
```
Comment (zromano) [inline src/main/java/com/smartsheet/api/internal/http/RequestAndResponseData.java]: nit: can we add a comment why we are doing `Math.max(0, headerValue.length() - 4)`?

File: src/main/java/com/smartsheet/api/internal/http/interceptor/HeadersInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+    public void setChangeAgent(String changeAgent) {
+        this.changeAgent = changeAgent;
+    }
+
+    public void setUserAgent(String userAgent) {
+        this.userAgent = userAgent;
+    }
+
+    @NotNull
```
Comment (zromano) [inline src/main/java/com/smartsheet/api/internal/http/interceptor/HeadersInterceptor.java:62]: nit: I'm not sure if we want to add jetbrains annotations. This project doesn't have Lombok annotations (yet at least) which is what I assume what you wanted here.

File: src/integration-test/java/WorkspaceResourcesIT.java (hunk 1 above)
Comment (apederson94) [reply src/integration-test/java/WorkspaceResourcesIT.java]: Done.

File: src/main/java/com/smartsheet/api/internal/WorkspaceResourcesImpl.java (hunk 2 above)
Comment (apederson94) [reply src/main/java/com/smartsheet/api/internal/WorkspaceResourcesImpl.java]: Done

File: src/main/java/com/smartsheet/api/internal/oauth/OAuthFlowImpl.java_859283488703776
Comment (apederson94) [reply src/main/java/com/smartsheet/api/internal/oauth/OAuthFlowImpl.java_859283488703776:1]: Yeah I have no idea where that came from 😂

File: src/main/java/com/smartsheet/api/internal/http/service/WorkspaceService.java (hunk 3 above)
Comment (apederson94) [reply src/main/java/com/smartsheet/api/internal/http/service/WorkspaceService.java:7]: I believe this was a copy/paste job by me

File: src/main/java/com/smartsheet/api/internal/http/RequestAndResponseData.java (hunk 4 above)
Comment (apederson94) [reply src/main/java/com/smartsheet/api/internal/http/RequestAndResponseData.java:315]: Good call. I reworked the logic into more digestible pieces

File: src/main/java/com/smartsheet/api/internal/http/RequestAndResponseData.java (hunk 5 above)
Comment (apederson94) [reply src/main/java/com/smartsheet/api/internal/http/RequestAndResponseData.java]: Done

File: src/main/java/com/smartsheet/api/internal/http/interceptor/HeadersInterceptor.java (hunk 6 above)
Comment (apederson94) [reply src/main/java/com/smartsheet/api/internal/http/interceptor/HeadersInterceptor.java:62]: This is actually something that is been enforced by the override of the intercept method for OkHttp, unfortunately. I'm not sure where that leave us, but yeah it is kind of outside of my control

**Pull request merged**